### DISCUSSION TITLE
different simulation result when applying the control rules between in inp file and in pyswmm

### DISCUSSION THREAD

**Discussion by Yunp-J** [pyswmm/pyswmm#461] [Q&A]
Hi, I'm trying to write some control rules in pyswmm, but only for the previous certain time during the simulation (e.g. 2-hour simulation, use new pyswmm controls rule for the first 1 hour, the remaining 1 hour use the existing rules in the inp file.). As I know, the priority of control rules of inp file is higher, so my plan is to write both 2 groups of rules in pyswmm and use a time variable to monitor when to switch the rules. However, when I test to rewrite the rules from the inp file in pyswmm, I found that the simulation results are a little bit different, and the differences happened after the control rules were applied.

for testing I only change the one following rule, and I delete this rule in the inp file: 
![image](https://github.com/pyswmm/pyswmm/assets/39219239/becc825d-1327-46b0-bfc7-d00d9382261c)
![image](https://github.com/pyswmm/pyswmm/assets/39219239/f40d2e2d-985f-4d16-9864-d5f559c8f831)

here is the result comparison (I use the copy of the inp file without deleting the rule to generate):
![image](https://github.com/pyswmm/pyswmm/assets/39219239/058764fb-cd3f-4603-a1ad-df9be3c0c73d)
![image](https://github.com/pyswmm/pyswmm/assets/39219239/f084c033-7e3b-4304-a9a4-91549f503b4d)

As you can see, at the beginning there are only 2 short peaks in the pyswmm plot, however, there are 3 in the swmm plot.

Besides, for testing, I use the equal simulation time step instead of the variated simulation time step. But this is not the solution. And in pyswmm I use the code **for step in sim:** and without **step_advance** to make sure using the same simulation step with inp.

Do you have any suggestions :)

**Comment by Yunp-J** [comment]
Besides, here is the definition for the 2 variables
![image](https://github.com/pyswmm/pyswmm/assets/39219239/44cb0908-c709-4ef2-8ce4-1ba229a281ba)

**Comment by bemcdonnell** [comment]
@Yunp-J Great discussion report!  Can you include any additional info on the simulation code (if you use `sim.step_advance()` )?  Those differences are subtle.  SWMM evaluates control rules on every routing time step.  PySWMM gives you the option (using step advance).  This is just a statement - you may not be using `step_advance`.

*** Maybe the problem ***
One note... Your SWMM rules converted to pyswmm are not truly the same.  the `elif` only allows entrance (like a C-switch statement).  Whereas, in your SWMM rules .. multiple rules could quality.  Therefore, you could emulate that with a rule Priority.   If you take the `el` off the `elif` and run it again it might work better.  Do you understand?

**Reply by Yunp-J** [reply]
```
with Simulation(new_measure_inp) as sim:
    #for rule JCW0000.1
    node_JCW0000 = Nodes(sim)["JCW0000"]
    weir_XQGBZ = Links(sim)["JCW0000.1"]   

    for step in sim:
        #rule JCW0000.1
        if node_JCW0000.head >= 18.2 and node_JCW0000.head < 18.4:
            weir_XQGBZ.target_setting = 0.5 
        if node_JCW0000.head >= 18.4:
            weir_XQGBZ.target_setting = 1         
        if node_JCW0000.head < 18.2:
            weir_XQGBZ.target_setting = 0    
```
Here is my code, and I didn't set the step_advance. Besides, I think you are right that I should take the el off as there is no priority in the SWMM rules. However, after I corrected my code, the result still doesn't change a lot, which means that the difference is still there :(

**Reply by bemcdonnell** [reply]
@Yunp-J try a routing step of 0.01 and see if you can get them closer.

**Reply by bemcdonnell** [reply]
@Yunp-J wait… where are you getting the timeseries data from for both those plots? The OUT files for both?

**Reply by Yunp-J** [reply]
no, for the pyswmm controller I got the plot from the OUT file, but for the SWMM controller I directly plot it in SWMM.

**Reply by bemcdonnell** [reply]
Ok they are both from the OUT files then

**Reply by bemcdonnell** [reply]
What SWMM versions are you comparing?  Check the RPT file headers

**Reply by Yunp-J** [reply]
v5.2.3

**Reply by Yunp-J** [reply]
and for pyswmm version is 1.4.0

**Comment by Yunp-J** [comment]
By the way, I changed the routing step of 0.01 and simulated again, but they still looked different :(
